### ISSUE TITLE
Semantic tokens features trackList

### ISSUE THREAD

**Issue by soulomoon** [haskell/haskell-language-server#3931]
Semantic heighlighting would gives us great experience on reading the code. It's been a shame that haskell does not have it for all these times. But now we are going to make it happen for haskell in hls. 
But since the amount of works it need, we would do them step by step. 
Here is the track list of features. 
* [x] Support semantic tokens type:  
  class and class method type family name (data family)
 data constructor name (not distinguishing record and normal data, and GADT)
 type constructor name (GADT)
 record field name
 type synonym
 pattern synonym
 functions
 none-function variables

* Support PluginMethodHandler
   - [x] [textDocument/semanticTokens/full](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#semanticTokens_fullRequest). done in [Semantic tokens](https://github.com/haskell/haskell-language-server/pull/3892)
   - [ ] [textDocument/semanticTokens/full/delta](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#semanticTokens_deltaRequest) 
   - [ ] [workspace/semanticTokens/refresh](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#semanticTokens_refreshRequest)
 * Support modifiers:
    - [ ] declaration (as in class declearations, type definition and type family)
    - [ ] definition (as in class instance declaration, left hand side value binding, and type family instance)
    - [ ] modification (as in rec field update)
* [x] Support unicode, done in [Semantic tokens](https://github.com/haskell/haskell-language-server/pull/3892)
* [x] Support imported name, done in [Semantic tokens](https://github.com/haskell/haskell-language-server/pull/3892)

**Comment by Bodigrim** [comment]
Thanks for working on this feature! I gave it a quick try in SublimeText on https://github.com/Bodigrim/cabal-add/blob/a83c827e45eeaed722d63f08a4c1bfd183770e97/src/Distribution/Client/Add.hs#L189:
<img width="372" alt="image" src="https://github.com/haskell/haskell-language-server/assets/2293557/156e7108-5dbb-4f54-ba0f-bc95bc91d0ff">
Is it something off with my configuration? Why `String` is in blue, but `ByteString` is white?

**Comment by michaelpj** [comment]
The current implementation highlights type synonyms differently to datatypes. So `String` and `FilePath` are different because they are type synonyms. Maybe this is confusing! (The fact that you asked suggests that it is...)

It's a bit funny that you get the same colour for `parseCabalFile` and `MonadError` also. Generally it seems a bit hit-and-miss whether any given colour scheme actually distinguishes everything that's in the spec...

**Comment by Bodigrim** [comment]
> Maybe this is confusing! (The fact that you asked suggests that it is...)

Yeah, that's very confusing, especially given how ubiquitous `String` and `FilePath` are. I think it would be better not to distinguish.

**Comment by soulomoon** [comment]
I think we may need a bit of setting in the color scheme to get the color right. The mapping of hs semantic token type to the standard semantic token type in lsp is here.

https://github.com/soulomoon/haskell-language-server/blob/master/plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Mappings.hs

As for if you want the color scheme to not distinguish between a normal datatype with a type synonyms. 
we can always set the `Type`( type synonym) and `Enum`(a normal datatype) to have the same color.

**Comment by soulomoon** [comment]
To be precise, we are not actually doing the semantic highlighting, we are doing giving semantic types to tokens that enables the semantic highlighting feature in the editors. Maybe it worth a note some where to help user to configure this if the color scheme does not look good. Expecially with the fact that there is a layer of mappings between hs semantic to lsp semantic.

**Comment by soulomoon** [comment]
I was thinking🤔， maybe we can make the `toLspTokenType` configurable to provide more flexibility on this.

**Comment by soulomoon** [comment]
Here is the the default mappings look like in vscode with the scheme `one dark pro`.
![image](https://github.com/haskell/haskell-language-server/assets/14073857/e46845d4-8944-40bd-9ed0-b93caf747835)

**Comment by michaelpj** [comment]
(Possibly we should split this off into another issue)

Worth pointing out that type synonyms are highlighted as "type" and in this case the difference is because _datatypes_ get a special highlighting.

> maybe we can also make the toLspTokenType configurable to provide more flexibility on this

Generally we try pretty hard to avoid configuration. It's usually much better if we can find something that's universally acceptable. 

I guess I could maybe be convinced to have settings for each thing where we have a choice of a specific or generic token type, e.g. "special highlighting for type synonyms: y/n", and if it's off then we use the generic one (e.g. "just a type"). We _could_ let people pick exactly the LSP semantic token type for each thing but that's probably overkill.

We do already have a configuration switch here: it's whether or not you have semantic tokens on or off. If you turn semantic tokens on, you should probably expect it to make some fine distinctions that aren't obvious to you at first sight. So a different question is "how do we make this discoverable?". Perhaps in this instance having the options would serve a double function: it tells you what kinds of things we distinguish, and then you can fiddle with it to see what it does.

Having written all that I guess I've convinced myself that we should have options. Especially since a lot of the things semantic tokens lets you do are quite opinionated and specific, which some people will inevitably dislike (e.g. I bet people will complain about the different highlighting of things of function type).

**Comment by michaelpj** [comment]
I split off another issue for discussing the configuration issue: https://github.com/haskell/haskell-language-server/issues/3937

**Comment by michaelpj** [comment]
@soulomoon a couple of minor things that we still need to do:
- Add you to CODEOWNERS for the plugin (I've invited you to the repository)
- Update the "Features" section of the documentation: https://github.com/haskell/haskell-language-server/blob/master/docs/features.md
    - New section for semantic tokens
    - New entry in the top table linking to it
    - Remove it from "unimplemented features"
- Update the "plugin support" section of the docs: https://github.com/haskell/haskell-language-server/blob/master/docs/features.md
    - I think it can be Tier 2
- Add it to the test job: https://github.com/haskell/haskell-language-server/blob/master/.github/workflows/test.yml

**Comment by soulomoon** [comment]
@michaelpj thanks for sort it out.

**Comment by michaelpj** [comment]
A possible concern: there is a `multilineTokenSupport` client capability. We probably need to respect that as well as `overlappingTokenSupport`. I think the current implementation might return multi-line tokens if something goes over several lines.

I think we could maybe fix this universally in `lsp`: the functions that compute semantic tokens could also take the `ClientCapabilities`, and then do something like:
- If `multilineTokenSupport` is false, then split any multi-line tokens into one token per line
- if `overlappingTokenSupport` is false, then split any overlapped tokens into `[<part before overlap>, <part after overlap>]`

**Comment by michaelpj** [comment]
I actaully don't know if we'll emit multi-line tokens today already, do you know @soulomoon ? I guess we probably don't, since all the stuff we recognize today can't be split over multiple lines anyway.

**Comment by michaelpj** [comment]
I made an issue for doing something helpful in lsp: https://github.com/haskell/lsp/issues/545

**Comment by MangoIV** [comment]
pretty cool! do you think it would be somehow possible to highlight parameters to a function? I could not figure out how that would work, i.e. that in
```haskell 
f = \x -> x y z
``` 
`x` gets a highlight group and that reoccurs in the body where `x` is in scope.

**Comment by MangoIV** [comment]
I've really enjoyed the possibility of distinguishing `class` and `type/ enum`. 
![image](https://github.com/haskell/haskell-language-server/assets/40720523/2dd41299-2c4a-4a40-8d72-c3bf6afd1546)

something that I'm wondering is why you chose to implement `class` and not `interface`, when I'm using the same colour schemes across languages, I think semantically it would make more sense if haskell `class`es would not be coloured like `java` classes but rather like `java` `interface`s or `abstract` classes or whatever.

**Comment by michaelpj** [comment]
> pretty cool! do you think it would be somehow possible to highlight parameters to a function?

There is a `paramter` token type. I'm unsure if that's supposed to be used for the declaration of a parameter or also its uses, since it means we can't mark it as a plan "variable". But I think it would be reasonable to use it for such things. We also have a bit of confusion because we use the `function` token type for variables of function type, so then we start to have difficulties: what token type do we give for a variable that is both a parameter and a function?

> something that I'm wondering is why you chose to implement class and not interface

I think either is plausible, I think @soulomoon just picked one. I guess on reflection `interface` is maybe a bit more accurate.

**Comment by soulomoon** [comment]
We won't have a perfect match on Haskell token vs default token. It actually configurable.

**Comment by MangoIV** [comment]
surely, yes, but I wonder what would happen if someone would program in both haskell and an object oriented language :D (I know even then you can configure it but perhaps we should optimize for the stock experience :P)

**Comment by soulomoon** [comment]
Yes, since haskell do have class term, it is a trade-off between prioritizing haskell term vs prioritizing object oriented language term. As a language server for haskell, I think the choice would be obvious?

**Comment by rumkeller** [comment]
@michaelpj
> There is a paramter token type. I'm unsure if that's supposed to be used for the declaration of a parameter or also its uses

[This](https://gist.github.com/swarn/fb37d9eefe1bc616c2a7e476c0bc0316#what-is-semantic-highlighting) seems to suggest that it's meant for both.

> what token type do we give for a variable that is both a parameter and a function?

[Here](https://gist.github.com/swarn/fb37d9eefe1bc616c2a7e476c0bc0316#dealing-with-ambiguity) is an example of how the text editor deals with multiple types for the same token. So perhaps it's o.k. to emit both variable and parameter types for the same token?

**Comment by soulomoon** [comment]
> So perhaps it's o.k. to emit both variable and parameter types for the same token
AFAIK, VSCode would consider them as overlapping token which they do not support.

**Comment by rumkeller** [comment]
You're right -- it [seems](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens) there can indeed be only one token type (though several modifiers). Apologies for the confusion.

So one has to choose between variable, function, and parameter. FWIW, here is what clangd does:
- Parameters are always marked as such, even if they are function pointers.
- It adds `functionScope` and `globalScope` modifiers. `functionScope` is for both parameters and local variables.

**Comment by michaelpj** [comment]
> It adds functionScope and globalScope modifiers. functionScope is for both parameters and local variables.

These are non-specified modifiers, right? Do any clients check for them? (Generally it's a bit sad that we don't have an easy way to agree on extensions to the list of modifiers...)

**Comment by rumkeller** [comment]
> These are non-specified modifiers, right? 

Indeed they're not mentioned in the [spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens).

> Do any clients check for them?  

Neovim exposes the modifiers verbatim to the color scheme -- quite easy to adjust, though the default scheme appears to not be aware of them.
